subsys/bluetooth/controller/coex/coex_ticker.c

```diff
@@ -151,59 +140,42 @@ static void coex_ticker_grant_irq_handler(const struct device *dev,
 					struct gpio_callback *cb, uint32_t pins)
 {
 	ARG_UNUSED(dev);
+	ARG_UNUSED(cb);
 	ARG_UNUSED(pins);
-	struct coex_ticker_data *data = CONTAINER_OF(cb, struct coex_ticker_data, grant_irq_cb);
-	const struct coex_ticker_config *config = data->dev->config;
 
 	LOG_DBG("COEX Inhibit IRQ Detected");
 	DEBUG_COEX_IRQ(1);
 	DEBUG_COEX_GRANT(0);
 
-	gpio_pin_interrupt_configure_dt(&config->grant_spec, GPIO_INT_DISABLE);
-	coex_ticker_grant_start(data->dev);
+	gpio_pin_interrupt_configure_dt(&grant_spec, GPIO_INT_DISABLE);
+	coex_ticker_grant_start();
 }
 
 
-static int coex_ticker_init(const struct device *dev)
+static int coex_ticker_init(void)
 {
-	const struct coex_ticker_config *config = dev->config;
-	struct coex_ticker_data *data = dev->data;
 	int res;
 
-	data->dev = dev;
-
 	DEBUG_COEX_INIT();
-	res = gpio_pin_configure_dt(&config->grant_spec, GPIO_INPUT);
+	res = gpio_pin_configure_dt(&grant_spec, GPIO_INPUT);
 	if (res) {
 		return res;
 	}
 
-	gpio_init_callback(&data->grant_irq_cb,
-					   coex_ticker_grant_irq_handler,
-					   BIT(config->grant_spec.pin));
+	gpio_init_callback(&grant_irq_cb, coex_ticker_grant_irq_handler,
+			   BIT(grant_spec.pin));
 
-	res = gpio_add_callback(config->grant_spec.port, &data->grant_irq_cb);
+	res = gpio_add_callback(grant_spec.port, &grant_irq_cb);
 	if (res) {
 		return res;
 	}
 
-	res = gpio_pin_interrupt_configure_dt(&config->grant_spec, GPIO_INT_EDGE_TO_INACTIVE);
+	res = gpio_pin_interrupt_configure_dt(&grant_spec, GPIO_INT_EDGE_TO_INACTIVE);
 	if (res) {
 		return res;
 	}
 
 	return 0;
 }
 
-#define COEX_NODE DT_PHANDLE(DT_NODELABEL(radio), coex)
-
-static struct coex_ticker_config config = {
-	.grant_spec = GPIO_DT_SPEC_GET(COEX_NODE, grant_gpios),
-	.grant_delay_us = DT_PROP(COEX_NODE, grant_delay_us)
-};
-static struct coex_ticker_data data;
-
-DEVICE_DEFINE(coex_ticker, "COEX_TICKER", &coex_ticker_init, NULL,
-	&data, &config,
-	APPLICATION, 90,
-	NULL);
+SYS_INIT(coex_ticker_init, APPLICATION, 90);
```

Comment:
Hey, I think this should rather become a `DEVICE_DT_DEFINE`: the driver may not make much use of the device structure, but it is based on a devicetree structure, if you instantiate it with `DEVICE_DT_DEFINE` then the corresponding struct device would get an ordinal, and the buildsystem will ensure that the init function get called after the corresponding device (the gpio controller in `grant-gpios`) have been already initialized.

Reply:
hmm good point, let me check how it's organized in dt